packages/metadata-editor/src/MetadataEditor.tsx

```diff
@@ -0,0 +1,325 @@
+        this.name,
+        JSON.stringify(newMetadata)
+      ).then((response: any): void => {
+        this.handleDirtyState(false);
+        this.onSave();
+        this.close();
+      });
+    }
+  }
+
+  handleTextInputChange(event: any, schemaField: string): void {
+    this.handleDirtyState(true);
+    // Special case because all metadata has a display name
+    if (schemaField == 'display_name') {
+      this.displayName = event.nativeEvent.srcElement.value;
+    } else {
+      this.metadata[schemaField] = event.nativeEvent.srcElement.value;
+    }
+  }
+
+  handleDropdownChange = (schemaField: string, value: string): void => {
+    this.handleDirtyState(true);
+    this.metadata[schemaField] = value;
+    this.update();
+  };
+
+  handleDirtyState(dirty: boolean): void {
+    this.dirty = dirty;
+    if (this.dirty && !this.title.className.includes(DIRTY_CLASS)) {
+      this.title.className += DIRTY_CLASS;
+    } else if (!this.dirty) {
+      this.title.className = this.title.className.replace(DIRTY_CLASS, '');
+    }
+  }
+
+  onUpdateRequest(msg: Message): void {
+    super.onUpdateRequest(msg);
+    // If the update request triggered rendering a 'code' input, and the editor hasn't
+    // been initialized yet, create the editor and attach it to the 'code' node
+    if (!this.editor && document.getElementById('code:' + this.id) != null) {
+      let initialCodeValue;
+      const getMimeTypeByLanguage = this.editorServices.mimeTypeService
+        .getMimeTypeByLanguage;
+      // If the file already exists, initialize the code editor with the existing code
+      if (this.name) {
+        initialCodeValue = this.metadata['code'].join('\n');
+      } else {
+        initialCodeValue = '';
+      }
+      this.editor = this.editorServices.factoryService.newInlineEditor({
+        host: document.getElementById('code:' + this.id),
+        model: new CodeEditor.Model({
+          value: initialCodeValue,
+          mimeType: getMimeTypeByLanguage({
+            name: this.metadata['language'],
+            codemirror_mode: this.metadata['language']
+          })
+        })
+      });
+      this.editor.model.value.changed.connect((args: any) => {
+        this.metadata['code'] = args.text.split('\n');
+        this.handleDirtyState(true);
+      });
+    }
+  }
+
+  getDefaultChoices(fieldName: string) {
+    let defaultChoices = this.schema[fieldName].uihints.default_choices;
+    if (defaultChoices == undefined) {
+      defaultChoices = [];
+    }
+    for (const otherMetadata of this.allMetadata) {
+      if (!defaultChoices.includes(otherMetadata.metadata[fieldName])) {
+        defaultChoices.push(otherMetadata.metadata[fieldName]);
+      }
+    }
+    return defaultChoices;
+  }
+
+  renderTextInput(
+    label: string,
+    description: string,
+    fieldName: string,
+    defaultValue: string
+  ) {
+    return (
+      <FormGroup
+        key={label}
+        label={label}
+        labelInfo="(required)"
```

Comment:
we should dynamically set this for every field based on the schema. Currently it's listed for all text inputs and drop down but not code. This is inaccurate since code is required for snippets but description (a text field) is not.